directory: Improve error message

keisku opened the pull request:
### What does this PR do?
<!-- A brief description of the change being made with this pull request. -->

Improve the error handling when Agent tries to read the not allowed directory.

### Motivation
<!-- What inspired you to submit this pull request? -->

#### `StopIteration` doesn't suggest next actions

Suppose we have the following directory and `conf.yaml`.

```
vagrant@vagrant:~$ whoami
vagrant
vagrant@vagrant:~$ mkdir -p test/foo
vagrant@vagrant:~$ mkdir -p test/bar
vagrant@vagrant:~$ chmod 700 test
vagrant@vagrant:~$ ls -la test
total 16
drwx------  4 vagrant vagrant 4096 Oct 12 04:38 .
drwxr-x--- 15 vagrant vagrant 4096 Oct 12 04:38 ..
drwxrwxr-x  2 vagrant vagrant 4096 Oct 12 04:38 bar
drwxrwxr-x  2 vagrant vagrant 4096 Oct 12 04:38 foo
```

```yaml
init_config:
instances:
  - directory: /home/vagrant/test
```

When we run this command,

```bash
sudo -u dd-agent datadog-agent check directory -l error
```

We have the non-actionable message.

```
    directory (1.13.1)
    ------------------
      Instance ID: directory:2e7e27b6b9a8ee26 [ERROR]
      Configuration Source: file:/etc/datadog-agent/conf.d/directory.d/conf.yaml
      Total Runs: 2
      Metric Samples: Last Run: 0, Total: 0
      Events: Last Run: 0, Total: 0
      Service Checks: Last Run: 1, Total: 2
      Average Execution Time : 5ms
      Last Execution Date : 2022-10-12 05:34:21 CEST / 2022-10-12 03:34:21 UTC (1665545661000)
      Last Successful Execution Date : Never
      Error:
      Traceback (most recent call last):
        File "/opt/datadog-agent/embedded/lib/python3.8/site-packages/datadog_checks/base/checks/base.py", line 1116, in run
          self.check(instance)
        File "/opt/datadog-agent/embedded/lib/python3.8/site-packages/datadog_checks/directory/directory.py", line 67, in check
          self._get_stats()
        File "/opt/datadog-agent/embedded/lib/python3.8/site-packages/datadog_checks/directory/directory.py", line 81, in _get_stats
          walker = [next(walker)]
      StopIteration
```

#### Hard to notice directories that are failing to scan

Suppose we have the following directories and `conf.yaml`,

```
vagrant@vagrant:~$ whoami
vagrant
vagrant@vagrant:~$ mkdir -p test/not_allowed
vagrant@vagrant:~$ chmod 700 test/not_allowed
vagrant@vagrant:~$ namei -mon test/not_allowed
f: test/not_allowed
 drwxrwxr-x vagrant vagrant test
 drwx------ vagrant vagrant not_allowed
```

```yaml
init_config:
instances:
  - directory: /home/vagrant/test
    recursive: true
```

We have no errors and exceptions.

```bash
sudo -u dd-agent datadog-agent check directory -l error
```

### Additional Notes
<!-- Anything else we should know when reviewing? -->

### Review checklist (to be filled by reviewers)

- [ ] Feature or bugfix MUST have appropriate tests (unit, integration, e2e)
- [ ] PR title must be written as a CHANGELOG entry [(see why)](https://github.com/DataDog/integrations-core/blob/master/CONTRIBUTING.md#pull-request-title)
- [ ] Files changes must correspond to the primary purpose of the PR as described in the title (small unrelated changes should have their own PR)
- [ ] PR must have `changelog/` and `integration/` labels attached
- [ ] If the PR doesn't need to be tested during QA, please add a `qa/skip-qa` label.

FlorentClarret reviewed (dismissed):
👍

alopezz commented on directory/datadog_checks/directory/directory.py:
```diff
@@ -78,7 +78,11 @@ def _get_stats(self):
-            walker = [next(walker)]
+            try:
+                walker = [next(walker)]
+            except Exception as e:
```
I'd rather avoid the naked exception here if we can.

alopezz commented on directory/datadog_checks/directory/directory.py:
```diff
@@ -78,7 +78,11 @@ def _get_stats(self):
+            try:
+                walker = [next(walker)]
+            except Exception as e:
+                self.log.error("failed to visit %s: %s", self._config.abs_directory, e)
```
```suggestion
                self.log.error("Failed to visit %s: %s", self._config.abs_directory, e)
```

alopezz commented on directory/datadog_checks/directory/traverse.py:
```diff
@@ -16,10 +16,7 @@ def _walk(top, follow_symlinks):
-        scandir_iter = scandir(top)
-    except OSError:
-        return
+    scandir_iter = scandir(top)
```
Why do we need to remove this try-except? Doesn't seem related to the new logging statement.

keisku commented:
@FlorentClarret @alopezz 
Thank you for your review!
https://github.com/DataDog/integrations-core/pull/13093/commits/0fce12d2d2e1ea231f4b0eecfd1c1d59d6cf9fbb didn't catch the exception when `recursive: true`. Now, moved `OSError` handling from `_walk()` to `_get_stats()`.

Moreover, I've updated "Motivation" of this PR description.

keisku commented on directory/datadog_checks/directory/traverse.py:
```diff
@@ -29,8 +26,6 @@ def _walk(top, follow_symlinks):
             entry = get_next(scandir_iter)
         except StopIteration:
             break
-        except OSError:
-            return
```
Same as https://github.com/DataDog/integrations-core/pull/13093/files#r994660105

keisku commented:
@alopezz Executed `ddev test directory -fs` to fix the code style! Thanks.
